mlx/fast.cpp

```diff
@@ -127,4 +127,168 @@ bool RoPE::is_equivalent(const Primitive& other) const {
+    std::ostringstream msg;
+    msg << "[scaled_dot_product_attention] query, keys expected to have matching last dimension; found query shape "
+        << queries.shape() << " for keys shape " << keys.shape() << ".";
+    throw std::invalid_argument(msg.str());
+  }
+
+  // head_dim must be 128.
+  constexpr const int supported_head_dim = 128;
+  if (queries.shape(-1) != supported_head_dim) {
+    std::ostringstream msg;
+    msg << "[scaled_dot_product_attention] query head dim must be 128 ; found "
+        << queries.shape(-1) << " .";
+    throw std::invalid_argument(msg.str());
+  }
+
+  // K, V must have matching number of heads (n_kv_heads);
+  size_t n_q_heads = queries.shape(-3);
+  size_t n_kv_heads = keys.shape(-3);
+
+  if (keys.shape(-3) != values.shape(-3)) {
+    std::ostringstream msg;
+    msg << "[scaled_dot_product_attention] keys, values expected to have matching n_kv_heads; found keys with n_heads "
+        << keys.shape(-3) << " for values with n_heads " << values.shape(-3)
+        << ".";
+    throw std::invalid_argument(msg.str());
+  }
+
+  // n_heads % n_kv_heads == 0; n_heads >= 1, n_kv_heads >= 1.
+  if (n_q_heads % n_kv_heads != 0) {
+    std::ostringstream msg;
+    msg << "[scaled_dot_product_attention] n_heads must be a multiple of n_kv_heads, found n_heads "
+        << n_q_heads << " for n_kv_heads " << n_kv_heads << ".";
+    throw std::invalid_argument(msg.str());
+  }
+
+  // q_seq_len == 1 for now in decoding (no mask) implementation
+  if (queries.shape(-2) > 1 && !mask.has_value()) {
+    std::ostringstream msg;
+    msg << "[scaled_dot_product_attention] query sequence length must be 1 at decode time, found  "
+        << queries.shape(-2) << ".";
+    throw std::invalid_argument(msg.str());
+  }
+
+  auto final_type = result_type({queries, keys, values});
+
+  auto q = astype(queries, final_type, s);
+  auto k = astype(keys, final_type, s);
+  auto v = astype(values, final_type, s);
+
+  auto out_shape =
+      std::vector<int>({q.shape(0), q.shape(1), q.shape(2), v.shape(-1)});
+
+  /* generic implementation for use cases that Metal implementation does not
+   * support. For non-supported cases listed below, use MLX primitives:
+   * * CPU implementation
+   * * batch size > 1
+   * * query sequence length > 1
+   * * non-null mask
+   */
+  bool needs_mask = mask.has_value();
+  auto fallback =
+      [scale, needs_mask, final_type, &s](const std::vector<array>& inputs) {
+        auto& q_tensor = inputs[0];
+        auto& k_tensor = inputs[1];
+        auto& v_tensor = inputs[2];
+        auto q_scaled = multiply(array(scale, q_tensor.dtype()), q_tensor, s);
+        int n_q_heads = q_tensor.shape()[1];
+        int n_kv_heads = k_tensor.shape()[1];
+
+        auto tile_if_needs_repeat = [](const array& arr,
+                                       int query_heads,
+                                       int kv_heads,
+                                       StreamOrDevice& s) -> array {
```

Comment:
nit: you could capture these in the lambda and type a little less in the function calls below.